cudamapper/src/main.cu

```diff
@@ -397,415 +182,341 @@ void align_overlaps(DefaultDeviceAllocator allocator,
+
+                // find anchors and overlaps
+                auto matcher = Matcher::create_matcher(device_allocator,
+                                                       *query_index,
+                                                       *target_index,
+                                                       cuda_stream);
+
+                std::vector<Overlap> overlaps;
+                OverlapperTriggered overlapper(device_allocator,
+                                               cuda_stream);
+                overlapper.get_overlaps(overlaps,
+                                        matcher->anchors(),
+                                        application_parameters.min_residues,
+                                        application_parameters.min_overlap_len,
+                                        application_parameters.min_bases_per_residue,
+                                        application_parameters.min_overlap_fraction);
+
+                // free up memory taken by matcher
+                matcher.reset(nullptr);
 
-    //Decrement counter which tracks number of overlap chunks to be filtered and printed
-    num_overlap_chunks_to_print--;
-};
+                // Align overlaps
+                std::vector<std::string> cigar;
+                if (application_parameters.alignment_engines > 0)
+                {
+                    cigar.resize(overlaps.size());
+                    CGA_NVTX_RANGE(profiler, "align_overlaps");
+                    align_overlaps(device_allocator,
+                                   overlaps,
+                                   *application_parameters.query_parser,
+                                   *application_parameters.target_parser,
+                                   application_parameters.alignment_engines,
+                                   cigar);
+                }
 
-} // namespace
+                // pass overlaps and cigars to writer thread
+                overlaps_and_cigars_to_write.add_new_element({std::move(overlaps), std::move(cigar)});
+            }
+        }
+    }
+}
 
-int main(int argc, char* argv[])
+/// \brief loads one batch into host memory and then processes its device batches one by one
+/// \param batch
+/// \param application_parameters
+/// \param host_cache data will be loaded into cache within the function
+/// \param device_cache data will be loaded into cache within the function
+/// \param overlaps_and_cigars_to_write overlaps and cigars are output here
+/// \param cuda_stream
+void process_one_batch(const BatchOfIndices& batch,
+                       const ApplicationParameters& application_parameters,
+                       DefaultDeviceAllocator device_allocator,
+                       IndexCacheHost& host_cache,
+                       IndexCacheDevice& device_cache,
+                       ThreadsafeProducerConsumer<OverlapsAndCigars>& overlaps_and_cigars_to_write,
```

Comment:
Optional: How about `overlaps_and_cigars_to_postprocess` instead?

Reply:
I went for `overlaps_and_cigars_to_process()` as at least for me that means both postprocess and write. What do you think?

Reply:
sounds good.